Development/boost/thread/win32_condition_variable.hpp

```diff
@@ -0,0 +1,444 @@
+// based on the <boost/thread/win32/condition_variable.hpp>
+
+#ifndef BOOST_THREAD_WIN32_CONDITION_VARIABLE_HPP
+#define BOOST_THREAD_WIN32_CONDITION_VARIABLE_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+# pragma once
+#endif // defined(_MSC_VER) && (_MSC_VER >= 1200)
+
+#include <boost/thread/win32/condition_variable.hpp>
+
+// Note: Windows `boost::condition_variable_any::wait` would throw nested lock exceptions when `boost::shared_mutex` has reached the
+// maximum number of exclusive_waiting locks. This problem occurrs inside the `boost::basic_condition_variable`'s do_wait_until(...),
+// when relocker is out scope. This could cause program termination due to unhandled exception.
+// This modified version of `condition_variable_any` presented below addresses the issues mentioned earlier.
+namespace boost
+{
+	namespace experimental
+    {
+        class basic_condition_variable
+        {
+            boost::mutex internal_mutex;
+            long total_count;
+            unsigned active_generation_count;
+
+            typedef boost::detail::basic_cv_list_entry list_entry;
+
+            typedef boost::intrusive_ptr<list_entry> entry_ptr;
+            typedef std::vector<entry_ptr> generation_list;
+
+            generation_list generations;
+            detail::win32::handle_manager wake_sem;
+
+            void wake_waiters(long count_to_wake)
+            {
+                detail::interlocked_write_release(&total_count, total_count - count_to_wake);
+                winapi::ReleaseSemaphore(wake_sem, count_to_wake, 0);
+            }
+
+            template<typename lock_type>
+            struct relocker
+            {
+                BOOST_THREAD_NO_COPYABLE(relocker)
+                    lock_type& _lock;
+                bool _unlocked;
+
+                relocker(lock_type& lock_) :
+                    _lock(lock_), _unlocked(false)
+                {
+                }
+                void unlock()
+                {
+                    if (!_unlocked)
+                    {
+                        try
+                        {
+                            _lock.unlock();
+                        }
+                        catch (...)
+                        {
+                            // ignore, threat this as unlocked
+                        }
+                        _unlocked = true;
+                    }
+                }
+                void lock()
+                {
+                    if (_unlocked)
+                    {
+                        _lock.lock();
+                        _unlocked = false;
+                    }
+                }
```

Comment:
Is there any possibility of `lock()` throwing?

Reply:
This is by design. The caller of the `wait` call can catch the throw and re-wait again.